libvmdk/libvmdk_handle.c

```diff
@@ -447,6 +447,150 @@ int libvmdk_handle_open(
 
 		goto on_error;
 	}
+
+  /*
+   * Extent filenames are not robust under file renaming, and
+   * monolithicSparse images are guaranteed to have an embedded descriptor
+   * file. Therefore, we use the image filename instead of the possibly
+   * stale extent filename for single-extent monolithicSparse images.
+   */
+  if ( internal_handle->descriptor_file->disk_type == LIBVMDK_DISK_TYPE_MONOLITHIC_SPARSE )
+  {
+    int extent_count = 0;
+    if ( libvmdk_descriptor_file_get_number_of_extents(
+          internal_handle->descriptor_file,
+          &extent_count, error) != 1 )
+    {
+      libcerror_error_set(
+		    error,
+		    LIBCERROR_ERROR_DOMAIN_RUNTIME,
+		    LIBCERROR_RUNTIME_ERROR_GET_FAILED,
+		    "%s: unable to retrieve number of extents.",
+		    function );
+
+		  goto on_error;
+    }
+
+    if ( extent_count == 1 )
+    {
+	    libvmdk_internal_extent_descriptor_t *extent_descriptor = NULL;
+      if ( libvmdk_descriptor_file_get_extent_by_index(
+		        internal_handle->descriptor_file,
+		        0,
+		        &extent_descriptor,
+		        error ) != 1 )
+		  {
+			  libcerror_error_set(
+	  	   error,
+		     LIBCERROR_ERROR_DOMAIN_RUNTIME,
+			   LIBCERROR_RUNTIME_ERROR_GET_FAILED,
+			   "%s: unable to retrieve extent: %d from descriptor file.",
+			   function,
+			   0 );
+
+			  goto on_error;
+		  }
+		  if ( extent_descriptor == NULL )
+  		{ 
+	  		libcerror_error_set(
+		  	 error,
+			   LIBCERROR_ERROR_DOMAIN_RUNTIME,
+			   LIBCERROR_RUNTIME_ERROR_VALUE_MISSING,
+  			 "%s: missing extent descriptor: %d.",
+	  		 function,
+		  	 0 );
+
+			  goto on_error;
+		  }
+
+      // now I have an extent_descriptor ho ho ho
+      if ( extent_descriptor->type == LIBVMDK_EXTENT_TYPE_SPARSE )
+      {
+
+        // need to copy filename into extent_descriptor->filename
+        // and set extent_descriptor->filename_size
+
+        memory_free(
+          extent_descriptor->filename );
```

Comment:
You'll need to set extent_descriptor->filename to NULL otherwise this can cause a double free.